x-pack/plugins/canvas/canvas_plugin_src/functions/types/common.ts

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Style } from './style';
+
+/**
+ * Utility type for converting a union of types into an intersection.
+ *
+ * This is a bit of "black magic" that will interpret a Union type as an Intersection
+ * type.  This is necessary in this case of distiguishing one collection from
+ * another in `FunctionError` and `FunctionStrings`.
+ */
+// prettier-ignore
+export type UnionToIntersection<U> = 
+  (U extends any ? (k: U) => void : never) extends ((k: infer I) => void) ? I : never;
+
+/**
+ * Utility type: gathers values of a collection as a type for use as a type.
+ */
+export type ValuesOf<T extends any[]> = T[number];
+
+/**
+ * Represents an object that is intended to be rendered.
+ */
+export interface Render<T> {
+  type: 'render';
+  as: string;
+  value: T;
+}
+
+/**
+ * Represents an object that is a Filter.
+ */
+export interface Filter {
+  type?: string;
+  value?: string;
+  column?: string;
+  and: Filter[];
+  to?: string;
+  from?: string;
+  query?: string | null;
+}
+
+/**
+ * Represents a function called by the `case` Function.
+ */
+export interface Case {
+  matches: any;
+  result: any;
+}
+
+// DATATABLES
+// ----------
+
+/**
+ * This type represents the `type` of any `DatatableColumn` in a `Datatable`.
+ */
+export type DatatableColumnType = 'string' | 'number' | 'boolean' | 'date' | 'null';
+
+/**
+ * This type represents the shape of a column in a `Datatable`.
+ */
+export interface DatatableColumn {
+  name: string;
+  type: DatatableColumnType;
+}
+
+/**
+ * A `Datatable` in Canvas is a unique structure that represents tabulated data.
+ */
+export interface Datatable {
+  columns: DatatableColumn[];
+  rows: any[];
```

Comment:
FWIW, the array needs to contain objects, and the objects should contain a `key: value` pair for each or the columns. For example:

```js
{
  type: 'datatable',
  columns: [{
    name: 'product',
    type: 'string',
  }, {
    name: 'price',
    type: 'number',
  }],
  rows: [{
    product: 'hat',
    price: 10.01
  }, {
    product: 'mat',
    price: 20.02
  }]
}
```

Not sure how hard that is to type. Even if you needed to keep it generic, wouldn't `Array<Record<string, any>>` be better?

Reply:
Very, very good to know.  I can update the type.

Reply:
Maybe it's a good place to ask, though likely not pertinent as a feedback item. Will it remain possible to add an alternative data table format eg. named `columntable` which would be like
```
{
  type: 'columndatatable',
  columns: [{
    name: 'product',
    type: 'string',
    values: ['hat', 'mat', ...]
  }, {
    name: 'price',
    type: 'number',
    values: [10.01, 20.02, ...]
  }]
}
```
or extend `datatable` to admit this alternative (as structurally it's quite similar)? In part, because of https://github.com/elastic/elasticsearch/pull/39287 - a feature which we suggested, and in part, because it'd be more efficient for a lot of uses and renderers? And then make
- functions - of the same name - that can work with this data format
- for functions that aren't (yet) supporting a columnar format, there'd be some automatic type (physical layout) conversion

I think the TS setup in this PR doesn't rule these out, in fact may make them more explicit when we get around to doing it?